presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java

```diff
@@ -117,6 +128,91 @@ public static ExpressionInterpreter expressionOptimizer(Expression expression, M
         return new ExpressionInterpreter(expression, metadata, session, expressionTypes, true);
     }
 
+    public static Object evaluateConstantExpression(Expression expression, Type expectedType, Metadata metadata, Session session)
+    {
+        ExpressionAnalyzer analyzer = ExpressionAnalyzer.createWithoutSubqueries(
+                metadata.getFunctionRegistry(),
+                metadata.getTypeManager(),
+                session,
+                EXPRESSION_NOT_CONSTANT,
+                "Constant expression cannot contain as sub-query");
+        analyzer.analyze(expression, new TupleDescriptor(), new AnalysisContext());
+
+        Type actualType = analyzer.getExpressionTypes().get(expression);
+        if (!canCoerce(actualType, expectedType)) {
+            throw new PrestoException(StandardErrorCode.INVALID_SESSION_PROPERTY, String.format("Can not set property of type %s to %s",
+                    expectedType.getTypeSignature().toString(),
+                    actualType.getTypeSignature().toString()));
+        }
+
+        IdentityHashMap<Expression, Type> coercions = new IdentityHashMap<>();
+        coercions.putAll(analyzer.getExpressionCoercions());
+        coercions.put(expression, expectedType);
+        return evaluateConstantExpression(expression, coercions, metadata, session);
+    }
+
+    public static Object evaluateConstantExpression(Expression expression, IdentityHashMap<Expression, Type> coercions, Metadata metadata, Session session)
+    {
+        // verify expression is constant
+        expression.accept(new DefaultTraversalVisitor<Void, Void>()
+        {
+            @Override
+            protected Void visitQualifiedNameReference(QualifiedNameReference node, Void context)
+            {
+                throw new SemanticException(EXPRESSION_NOT_CONSTANT, expression, "Constant expression cannot contain column references");
+            }
+
+            @Override
+            protected Void visitInputReference(InputReference node, Void context)
+            {
+                throw new SemanticException(EXPRESSION_NOT_CONSTANT, expression, "Constant expression cannot contain input references");
```

Comment:
How can this occur? The phrase "input references" doesn't mean anything to a user